### PR TITLE
PERF: add Numba engine for unifrac in beta_diversity

### PR DESCRIPTION
**Pull request by l1joseph**
Adds `engine="numba"` support to `beta_diversity` for both `unweighted_unifrac` and `weighted_unifrac`, alongside the existing default path. Both metrics currently go through `scipy.spatial.distance.pdist` with a Python callable invoked once per pair, so for n samples that's O(n^2) Python function calls. The numba kernels compute the whole condensed distance matrix in one parallel pass instead. Weighted covers both the normalized and unnormalized variants in a single kernel. The fast path only applies when the caller hasn't supplied a custom `pairwise_func` or any metric kwargs the numba kernels don't consume; otherwise it falls back to the existing cython/`pairwise_func` path exactly as before, with a warning if `engine="numba"` was explicitly requested.

Both kernels parallelize over `prange(n_half)` with a mirror-row pairing (row `i` paired with `n_samples - i - 2`) instead of a plain `prange(n_samples - 1)`. Row `i` has `n_samples - 1 - i` pairs to compute, so the naive form gives row 0's thread far more work than the last row's thread. Pairing each row with its mirror keeps the work per `prange` iteration roughly constant regardless of `i`.

Benchmark, `beta_diversity(engine="numba")` vs the default path, random tree and OTU table:

| samples x tips | cython | numba | speedup |
|---|---|---|---|
| 20 x 100 | 1.24ms / 2.03ms | 0.43ms / 0.44ms | unweighted 2.9x, weighted 4.6x |
| 50 x 500 | 10.8ms / 17.9ms | 2.9ms / 2.3ms | unweighted 3.7x, weighted 7.7x |
| 100 x 1000 | 59.0ms / 96.2ms | 14.4ms / 8.8ms | unweighted 4.1x, weighted 11.0x |

Numba JIT warm up is a one time cost on first call, and isn't part of the numbers above.

Tests:
- numba vs cython parity on the existing fixture, for both metrics and both weighted variants
- a larger random case with a seeded rng, including an all zero row
- a single sample degenerate size edge case
- invalid engine raises ValueError
- a custom `pairwise_func` is actually used even when `engine="numba"` is also requested
- an unrecognized kwarg still raises with `engine="numba"`, matching the cython path

Checklist:
- [x] I have read the contribution guidelines.
- [x] I have documented all public-facing changes in the changelog.
- [x] This pull request does not include code, documentation, or other content derived from external source(s).

### CODE REVIEW

**Comment by sfiligoi** [comment]
@l1joseph Can you provide benchmarks for larger inputs?
Something that would take at least 10s of seconds to run? Maybe 25k tips?

Please also disclose the CPU you tested this on (e.g. number of cores)

**Comment by l1joseph** [comment]
@sfiligoi Benchmarks at n=300, tips=25000, run on a Cosmos node (MI300A, 192 cores):

| metric | cython | numba | speedup |
|---|---|---|---|
| unweighted_unifrac | 14.9s | 1.94s | 7.65x |
| weighted_unifrac | 15.9s | 1.25s | 12.73x |

**Review by sfiligoi — APPROVED** [review]
Looks good to me

**Review by qiyunzhu — APPROVED** [review]
Looks good to me too. The parallelization strategy (pairing row i with mirror_i = n_samples - i - 2) looks smart to me. Previously I was trying to chunk rows and use OpenMP dynamic scheduling. I think the strategy in this PR may not be optimal for memory efficiency, because it has to jump from to top to the bottom of the matrix in each iteration. But perhaps my concern has already been prove unnecessary by benchmarks.

File: skbio/diversity/beta/tests/test_unifrac.py [hunk 1]
```diff
@@ -686,6 +688,146 @@ def test_weighted_unifrac_pycogent_adapted(self):
...
+        np.testing.assert_allclose(
+            dm_nb.data, dm_cy.data, rtol=1e-12, atol=1e-12)
+
+    @numba_code
+    def test_unweighted_unifrac_engine_numba_larger_random(self):
+        rng = np.random.default_rng(0)
+        counts = rng.integers(0, 10, size=(12, 5))
+        # cover the "both samples empty" (observed == 0) branch
+        counts[0] = 0
+        ids = [f"S{i}" for i in range(counts.shape[0])]
```
[the remark this answers is not in the document]
Comment (sfiligoi) [reply skbio/diversity/beta/tests/test_unifrac.py:713]: This looks completely irrelevant to me.

@l1joseph Can you please confirm that Copilot completely misunderstood the test?

**Comment by l1joseph** [comment]
@qiyunzhu Total memory traffic is the same either way, every row gets read exactly once regardless of which thread owns it, mirroring only changes the assignment. The jump from row i to mirror_i happens once per iteration, not repeatedly, so it's a one-time transition rather than thrashing. Without mirroring, plain `prange(n_samples - 1)` gives severe load imbalance since each row's inner loop does tree traversal, not just arithmetic, so an early thread doing O(n) work while a late thread does O(1) is the bigger cost here. Numba's `prange` also doesn't have OpenMP's `schedule(dynamic)`.

Also pushed a fix for the Copilot-flagged test: `test_unweighted_unifrac_engine_numba_larger_random` claimed to cover the both-samples-empty branch but only forced one all-zero row, so with seed 0 it never actually hit that case. Matched the weighted-unifrac test's pattern (both rows 0 and 1 forced to zero, explicit assertion on the (0,1) pair). 38/38 pass.

**Comment by sfiligoi** [comment]
Indeed, keeping the parallel ranges constant across all loops is essential to get good performance. Paring first and last is one of the best strategies I am aware of. (used in other functions, too)

**Comment by qiyunzhu** [comment]
@l1joseph @sfiligoi Good to know. Thanks for the knowledge!
